node/pkg/watchers/evm/ccq_backfill.go

```diff
@@ -0,0 +1,303 @@
+package evm
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/certusone/wormhole/node/pkg/watchers/evm/connectors"
+
+	"go.uber.org/zap"
+
+	ethHexUtil "github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/rpc"
+	ethRpc "github.com/ethereum/go-ethereum/rpc"
+)
+
+const CCQ_MAX_BATCH_SIZE = int64(1000)
+const CCQ_TIMESTAMP_RANGE_IN_MS = uint64(30 * 60)
+
+type ccqBackfillRequest struct {
+	timestamp uint64
+}
+
+// ccqRequestBackfill submits a request to backfill a gap in the timestamp cache.
+func (w *Watcher) ccqRequestBackfill(timestamp uint64) {
+	select {
+	case w.ccqBackfillChannel <- &ccqBackfillRequest{timestamp: timestamp}:
+		w.ccqLogger.Debug("published backfill request", zap.Uint64("timestamp", timestamp))
+	default:
+		// This will get retried next interval.
+		w.ccqLogger.Error("failed to published query response error to handler", zap.Uint64("timestamp", timestamp))
+	}
+}
+
+// ccqBackfiller is the runnable that handles backfilling the timestamp cache. It initializes the cache and then processes incoming backfill requests from CCQ.
+func (w *Watcher) ccqBackfiller(ctx context.Context) error {
+	if err := w.ccqBackfillInit(ctx); err != nil {
+		return fmt.Errorf("failed to backfill timestamp cache: %w", err)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case evt := <-w.ccqBackfillChannel:
+			w.ccqPerformBackfill(ctx, evt)
+		}
+	}
+}
+
+type (
+	ccqBatchResult struct {
+		result ccqBlockMarshaller
+		err    error
+	}
+
+	ccqBlockMarshaller struct {
+		Number ethHexUtil.Uint64
+		Time   ethHexUtil.Uint64 `json:"timestamp"`
+		// Hash   ethCommon.Hash    `json:"hash"`
+	}
+)
+
+// ccqBackfillInit determines the maximum batch size to be used for backfilling the cache. It also loads the initial batch of timestamps.
+func (w *Watcher) ccqBackfillInit(ctx context.Context) error {
+	// Get the latest block so we can use that as the starting point in our cache.
+	latestBlock, err := connectors.GetLatestBlock(ctx, w.ccqLogger, w.ethConn)
+	if err != nil {
+		return fmt.Errorf("failed to look up latest block: %w", err)
+	}
+	latestBlockNum := int64(latestBlock.Number.Uint64())
+	w.ccqLogger.Info("looked up latest block", zap.Int64("latestBlockNum", latestBlockNum), zap.Uint64("timestamp", latestBlock.Time))
+
+	var blocks Blocks
+	if w.ccqBatchSize == 0 {
+		// Determine the max supported batch size and get the first batch which will start with the latest block and go backwards.
+		var err error
+		w.ccqBatchSize, blocks, err = ccqBackFillDetermineMaxBatchSize(ctx, w.ccqLogger, w.ethConn, latestBlockNum, 100*time.Millisecond)
+		if err != nil {
+			return fmt.Errorf("failed to determine max batch size: %w", err)
+		}
+	} else {
+		blocks = append(blocks, Block{BlockNum: latestBlock.Number.Uint64(), Timestamp: latestBlock.Time})
+		w.ccqLogger.Info("using existing batch size for timestamp cache", zap.Int64("batchSize", w.ccqBatchSize))
+	}
+
+	if len(blocks) == 0 {
```

Comment:
Is this even possible?  Wouldn't you have, at least, the latest block in here, if nothing else?

Reply:
It shouldn't but the for loop would panic if it did, so I would like to keep it. I added a comment.